sensorprocess/sensorprocess.go

```diff
@@ -50,35 +68,46 @@ func Start(
 }
 
 // addLidarReading adds a lidar reading to the cartofacade.
-func addLidarReading(
-	ctx context.Context,
-	config Config,
-) bool {
-	tsr, err := config.Lidar.TimedLidarSensorReading(ctx)
-	if err != nil {
-		config.Logger.Warn(err)
-		// only end the sensor process if we are in offline mode
-		if config.LidarDataRateMsec == 0 {
-			return strings.Contains(err.Error(), replaypcd.ErrEndOfDataset.Error())
-		}
-		return false
-	}
+func (config *Config) addLidarReading(ctx context.Context) bool {
 	/*
-	 when the data rate msec is 0, we assume the user wants to be in "offline"
-	 mode and ensure every scan gets processed by cartographer
+	 when the lidar data rate msec is non-zero, we assume the user wants to be in "online"
+	 mode and ensure the most recent scan gets processed by cartographer. If data rate msec
+	 is zero we process every scan in order
 	*/
-	if config.LidarDataRateMsec == 0 {
-		tryAddLidarReadingUntilSuccess(ctx, tsr.Reading, tsr.ReadingTime, config)
-	} else {
-		timeToSleep := tryAddLidarReading(ctx, tsr.Reading, tsr.ReadingTime, config)
+	if config.LidarDataRateMsec != 0 {
+		// get next lidar data response
+		tsr, status, err := getTimedLidarSensorReading(ctx, config)
+		if err != nil {
+			return status
+		}
+
+		// sleep remainder of time interval
+		timeToSleep := tryAddLidarReading(ctx, tsr.Reading, tsr.ReadingTime, *config)
 		time.Sleep(time.Duration(timeToSleep) * time.Millisecond)
 		config.Logger.Debugf("sleep for %s milliseconds", time.Duration(timeToSleep))
+	} else {
+		// add the stored lidar data and begin processing the next one, if no imu exists or the currently stored imu data occurs after it.
+		if config.IMUName == "" || config.nextData.lidarTime.Sub(config.nextData.imuTime).Milliseconds() <= 0 {
+			if config.nextData.lidarData != nil {
+				tryAddLidarReadingUntilSuccess(ctx, config.nextData.lidarData, config.nextData.lidarTime, *config)
+				config.started = true
```

Comment:
What `started` does, is only allows imu data to be sent to cartographer after the first lidar data is added. If started is not set, it will continue to poll for more imu data until one is found after the first lidar reading. This allows that lidar reading to be added, `started` to be set to true, and imu data to begin being added.

Similar to stopped, we only want to add imu data when there is lidar data available for the constraint equations to work off of. Without this, I experienced issues where saved replay data was gathered from the movement sensors several seconds in advance of lidar data leading to ~60 imu measurements being added to carto, (and failing) before the first lidar measurement could occur.